src/bounds.py

```diff
@@ -5,18 +5,58 @@
 a list of features, here we only concern ourselves with lists of integers.
 """
 
-
-def lower_bound(x: list[int], v: int) -> int:
+def lower_bound(x: list[int], v: int, low: int=0, high: bool=None) -> int:
     """Get the index of the lower bound of v in x.
+    
 
     If all values in x are smaller than v, return len(x).
+    >>> lower_bound([2,3,5,5,7], 1)
+    0
+    >>> lower_bound([2,3,5,5,7], 2)
+    0
+    >>> lower_bound([2,3,5,5,7], 5)
+    2
+    >>> lower_bound([2,3,5,5,7], 8)
+    5
+    >>> lower_bound([2,3,5,5,7], 6)
+    4
+    >>> lower_bound([2,3,5,5,7], 4)
+    2
     """
-    return 0  # FIXME: Obviously the answer isn't always 0
-
+    high = high or len(x)
+    if low >= high: return len(x) # if all elements in x are smaller than v. 
+    mid = (low+high)//2 
+    if mid == 0 and x[mid] >= v:
+        return mid
+    elif x[mid] >= v and x[mid-1] < v:
+        return mid
+    elif x[mid] > v:
+        return lower_bound(x, v, low, mid)
+    else: # x[mid] < v
+        return lower_bound(x, v, mid+1, high)
```

Comment:
As far as I can tell, this works, but you handle a lot more cases than necessary. With the invariant you have, the general case is really enough. If we say that we are looking for the midpoint where we switch from below `v` to above or equal to `v` (with the twist that "switch" could be the first index if it is already above or equal), then we just need to ensure that we do not move `low` above the switch point or `high` below.

![lower-bound](https://user-images.githubusercontent.com/3206538/200115191-b10f1fac-8a9b-462f-8a23-0b29cbe8fc56.png)

If `mid` is less than `v`, then we can safely move `low` up to `mid+1` (it might move up to the _first_ point that is equal to `v`, but it can't move above it if `mid` is smaller). If mid is higher than `v`, then we can safely move `high` down there and keep the invariant true. There is no need to check things like `mid-1` or such; the general case will shrink the interval while keeping the switch point--the lower bound--inside it.

```python
def lower_bound(x: list[int], v: int) -> int:
    lo, hi = 0, len(x)
    while lo < hi:
        m = (lo + hi) // 2
        if x[m] < v:
            lo = m + 1
        else:
            hi = m
    return lo
```

Reply:
Hej Thomas, jeg forstår ikke helt, hvordan invarianten kan virke ift. high. high må vel heller ikke rykkes ned til switch point, må den? Som i fx x = [1,2,3,5,6,7] og v = 4. Her vil x[mid]=5 > v, men hvis vi så sætter high=mid, så vil lower bound (5) jo ikke længere være en del af det interval, der søges i.

Reply:
Invarianten siger at $x[high] >= v$, ikke $x[high] > v$, så den må gerne rykke ned hvis du kikker på `v`. Den må ikke rykke _under_ switch point, men den må gerne rykke ned til den (og så vil `low` senere hoppe der op også).

Du har ret i at i dit eksempel så vil `v` ikke længere være i intervallet, men det var den aldrig til at begynde med!

Nedenfor vil jeg skrive `-` for værdier vi ved er mindre end `v` (det er alle under `low`) og `+` for alle vi ved er større eller lig (alle lig eller over `high`).

Vi kører

```
[1, 2, 3, 5, 6, 7]
 ^        ^       ^
low      mid      high
```

hvor `x[mid] > 4` så vi ved at alle fra dette punkt og op er større lig (`+`)


```
[1, 2, 3, +, +, +]
 ^        ^       
low      high
```

Vi ved ikke lige nu om `high` peger højere end lower bound, den kan pege _på_ lower bound, men vi ved at lower bound ikke ligger højere end `high`.

Nu er `x[mid] == 2 < 4` så vi ved at alle indeks op til, og inklusiv, `mid` er for små, så de ryger i kategorien `-`

```
[-, -, 3, +, +, +]
       ^  ^       
   low   high
```

`low` er det første sted som lower bound kan ligge på (hvis vi tjekker værdien vil vi opdage at den er for lille, men det er i følge invarianten det først sted den kan ligge), og `high` er det sidste sted den kan ligge.

Nu bliver `mid = low`, den har værdi `x[mid] = 3 < 4`, så vi rykker `low` op til `mid+1`, og eliminerer derved indgang `mid` som kandidat. Nu står både `low` og `high` på det samme indeks. Vi ved at alle indeks under `low` er for små og at alle indeks fra `high` (inklusiv) og op er større lig, så vi ved at `low` må være lower bound.

Det at `high` rykkede ned da vi så på 5 er ikke et problem. Vi eliminrer ikke som kandidat den værdi `high` peger på her. (Det gør vi med en normal binær søgning, men ikke med den invariant vi har). Invarianten giver os at `high` er det højeste punkt hvor lower bound kan ligge. High selv er kandidat da den ligger på den anden side af switch point, men alle der ligger til højer for dem er så elimineret--de ligger til højre for switch point men de er ikke de mindste der gør det. Om ikke andet er `high` mindre end dem.

Så tænker vi i switch point fra under `v` (klasse `-`) til over-lig `v` (klasse `+`), så er alle _under_ `low` i klasse `-` og alle over eller lig `high` i klasse `+`. Hvis `low == high` så er alle under dem under switch point, alle over dem over switch point, så de må selv pege på switch point.